### ISSUE TITLE
There's no easy way to specify authentication

### ISSUE THREAD

**Issue by Jericho** [Pathoschild/FluentHttpClient#10]
The readme says that I can create my own implementation of IHttpFilter for authentication but this means that every developer using FluentHttpClient must "reinvent the wheel" for this simple task. Since most API require some form of authentication, I believe FluentHttpClient should make it easier to provide authentication.

I propose the following:

#### We should be able to configure authentication for each request
1. I suggest adding `IRequest WithAuthentication(string scheme, string value);` to IRequest (and Request.cs of course).
2. Add convenient methods for the most common authentication schemes. The two that I have most frequently seen are: 'bearer' (which is when you provide an api key) and 'basic' (which is when you provide a username and password). These two methods would be called `WithBearerAuthentication` and `WithBasicAuthentication` and would simply invoke the `WithAuthentication` method with appropriate parameters.

#### We should also be able to configure authentication for the client
3. I suggest adding a new IHttpFilter called `AuthorizationFilter`. This filter would automatically add the authentication to every request
4. I suggest adding `public void AddAuthentication(string scheme, string value);` to IClient (and FluentClient.cs of course). This method would instantiate a new `AuthenticationFilter` with provided parameters and add it to the list of filters.
5. Add convenient methods for the most common authentication schemes. As I previously mentioned, the two that I think are the most common are 'bearer' and 'basic'. These two methods would be called `AddBearerAuthentication` and `AddBasicAuthentication` and would simply invoke the `AddAuthentication` method with appropriate parameters.

I will be away for a while but I volunteer to submit a PR to add this functionality when I return.

**Comment by Pathoschild** [comment]
Your suggestions would work, but I want to keep the main fluent interface as simple as possible (e.g. I wouldn't want to add four methods to support four auth schemes). How about...

* Add an overload like `.Header(IFluentHeader)` to support many authorization schemes without adding an equal number of methods:
   ```cs
      .WithHeader(key, value)
      .WithHeader(new AuthHeader(scheme, value))
      .WithHeader(new BearerAuthHeader(token))
      .WithHeader(new BasicAuthHeader(user, password))
   ```
* Which would let us create one filter for any default header:
   ```cs
   // default custom header
   client.Filters.Add(new DefaultHeaderFilter(key, value));

   // default IFluentHeader
   client.Filters.Add(new DefaultHeaderFilter(authHeader));
   ```
* And would also make it easy to support default non-auth headers:
   ```cs
      .WithHeader(CacheControlHeader.MaxAge(60))
   ```

**Comment by Jericho** [comment]
I understand and agree with your desire to keep the interfaces as simple as possible. What I'm suggesting is in line with this desire: my suggestion is to add one single method to the 'client' interface and one method to the 'request' interface. The other methods I mentioned would be extension methods that would simply invoke the new interface methods.

Another important aspect is discoverability, particularly since authentication is necessary for interacting with almost every single API. We must make sure the means of providing authentication information is easily discoverable by developers. Along these lines I feel that `myFluentClient.AddAuthentication(...)` is easier to figure out than something like `myFluentClient.AddHeader(new SomeAuthClass(...))`. Intellisence in Visual Studio will also help us: developers will type the first few letters of the word "authentication" and will be presented with the new method.

Anyway, long story short: let me submit my PR with what I have in mind and we can review it and discuss it. I will not be offended if ultimately you decide not to accept my submission if you feel I am complicating the interface.

p.s.: I will be away for a few days so it might be a week or so before I submit all the PRs I have recently promised!

**Comment by Pathoschild** [comment]
Sure, submit the PR. :)

**Comment by Pathoschild** [comment]
Done in `develop` for the upcoming 2.4 release via PR #15. Thanks!
